plugins/engines/ospray/ispc/render/SimulationRenderer.ispc

```diff
@@ -106,308 +106,253 @@ inline void composite(const varying vec4f& src, varying vec4f& dst,
+                dg.epsilon = epsilon;
+                dg.Ns = gradient;
+
+                gradient =
+                    safe_normalize(volume->computeGradient(volume, point));
+
+                // Shading according to computed normal
+                const vec2f s = make_vec2f(0.5f);
+                const Light_SampleRes light =
+                    self->abstract.lights[0]->sample(self->abstract.lights[0],
+                                                     dg, s);
+                const float cosNL =
+                    (gradient.x == 0.f && gradient.y == 0.f &&
+                     gradient.z == 0.f)
+                        ? 1.f
+                        : abs(dot(safe_normalize(light.dir), gradient));
+
+                volumeSampleColor = volumeSampleColor * cosNL * light.weight;
+
+                // Specular
+                const vec3f reflectedNormal = normalize(
+                    ray.dir - 2.f * dot(ray.dir, gradient) * gradient);
+                const float angle =
+                    powf(max(0.f, dot(light.dir, reflectedNormal)), 20.f);
+                volumeSampleColor =
+                    volumeSampleColor + volume->specular * angle;
+            }
+
+            if (samplingCounter == 0 && self->shadows > 0.f)
+                // Compute shadow contributions
+                shadowIntensity =
+                    getVolumeShadowContributions(volume, self, ray, sample,
+                                                 point, epsilon);
         }
+        composite(make_vec4f(volumeSampleColor, sampleOpacity), pathColor,
+                  2.0f);
+
+        ++samplingCounter;
+    }
+
+    // Apply shadow to RGB values only
+    return pathColor * make_vec4f(make_vec3f(1.f - shadowIntensity), 1.f);
+}
+
+inline float shadedLightIntensity(varying ScreenSample& sample,
+                                  const varying Ray& ray,
+                                  const ShadingAttributes& attributes,
+                                  const varying vec3f& lightDirection,
+                                  DifferentialGeometry& dg)
+{
+    float shadowIntensity = 0.f;
+
+    vec3f ld = lightDirection;
+    if (attributes.self->softShadows != 0.f)
+        // Slightly alter light direction for Soft shadows
+        ld = normalize(ld +
+                       attributes.self->softShadows *
+                           getRandomVector(attributes.self->abstract.super.fb,
+                                           sample, attributes.normal,
+                                           attributes.self->randomNumber));
+
+    Ray shadowRay = ray;
+    setRay(shadowRay, dg.P, ld);
+    shadowRay.t0 = dg.epsilon;
+    shadowRay.t = inf;
+    shadowRay.time = sample.ray.time;
+    shadowRay.geomID = -1;
+
+    while (shadowIntensity < 1.f)
+    {
+        traceRay(attributes.self->abstract.super.model, shadowRay);
+
+        if (shadowRay.geomID == -1)
+            break;
+
+        DifferentialGeometry dg2;
```

Comment:
Is there a better name than dg2?

Reply:
Yep, shadowingGeometry is the correct term :)